### PR TITLE
Rewrite MinifyLocals from Uglify to Acorn. NFC

### PR DESCRIPTION
**Pull request by kripken**
Basically a simple rewrite from one AST to another, but also using slightly
more modern JS while doing so (`Set`, `Map`). Some parts differ though as
the ASTs are not compatible, so the algorithm is not 100% identical, but the
results should be.

Test changes are just minor whitespace, caused by the different whitespace
style that Uglify used to emit vs the new one.

Verified on `wasm2js2, 3, s, z` which we do not run on CI here.

After this we have just one remaining pass using Uglify, MinifyGlobals, after
which we can remove it entirely.

cc @walkingeyerobot

### CODE REVIEW

**Comment by sbc100** [comment]
Is this part of https://github.com/emscripten-core/emscripten/issues/12410?

File: tools/acorn-optimizer.js [hunk 1]
```diff
@@ -1444,9 +1606,11 @@ var registry = {
   applyDCEGraphRemovals: applyDCEGraphRemovals,
   minifyWhitespace: function() { minifyWhitespace = true },
   noPrint: function() { noPrint = true },
+  last: function() {},
```
Comment (sbc100) [inline tools/acorn-optimizer.js:1609]: What is this?

File: tools/acorn-optimizer.js [hunk 2]
```diff
@@ -1334,6 +1334,168 @@ function safeHeap(ast) {
...
+// Name minification
+
+var RESERVED = new Set(['do', 'if', 'in', 'for', 'new', 'try', 'var', 'env', 'let', 'case', 'else', 'enum', 'void', 'this', 'void', 'with']);
+var VALID_MIN_INITS = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_$';
+var VALID_MIN_LATERS = VALID_MIN_INITS + '0123456789';
+
+var minifiedNames = [];
+var minifiedState = [0];
+
+function ensureMinifiedNames(n) { // make sure the nth index in minifiedNames exists. done 100% deterministically
```
Comment (sbc100) [inline tools/acorn-optimizer.js:1346]: Can you put the comment on its own line (before the function I guess?)

**Review by sbc100 — APPROVED** [review]
Great!

I only briefly glanced over the actual algorithm, assuming that its its a roughly 1-to-1 with the existing code.

**Comment by kripken** [comment]
> Is this part of #12410?

It could be perhaps as part of a larger refactoring. But the Acorn code uses Terser for its output, so removing Uglify would not affect our usage of Terser.

File: tools/acorn-optimizer.js (hunk 1 above)
Comment (kripken) [reply tools/acorn-optimizer.js:1609]: The python code uses "last" as a marker for knowing that nothing will run next. The acorn code doesn't actually need it, as you can see here, but adding this lets the existing python driver code work unmodified. Basically this makes the new pass implement the exact same API surface as the old pass.

Comment (kripken) [reply tools/acorn-optimizer.js:1609]: (A refactoring on the python side could remove it, i'll add a TODO)

File: tools/acorn-optimizer.js (hunk 2 above)
Comment (kripken) [reply tools/acorn-optimizer.js]: Done.

**Pull request merged**